File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func hasCycle(_ head: ListNode?) -> Bool {
        // Solution 1: Fast and Slow pointers (Floyd's Tortoise and Hare Problem)
        var slow = head
        var fast = head?.next

        if slow == nil || fast == nil {
            return false
        }

        while slow != nil {
            if slow?.val == fast?.val {
                return true
            }
            slow = slow?.next
            fast = fast?.next?.next
        }
        
        return false

        //----------------------------------------------------------------------

        // Solution 2: Using a Set
        
    }
}

File: Data Structures & Algorithms/palindrome-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        mid = ListNode(slow.val)
        
        prev = None
        while slow != None: # want slow to overflow so prev is last
            nxt = slow.next
            slow.next = prev
            prev = slow
            slow = nxt
        
        l = head
        r = prev
        while r != None:
            if l.val != r.val:
                return False
            l = l.next
            r = r.next

        return True